examples/10.dynamic_configuration/README.md

```diff
@@ -0,0 +1,125 @@
+
+And finally the example shows how a separate sandpit compartment can be used by a subscriber to safely validate new data and gain trust in its content without exposing it own stack and heap.
+
+## Compartments in the example
+
+### Config Source (Publisher)
+The **Config Source** compartment provides the publisher of configuration data. In a real system this would receive updates from the network, although its possible that some compartments might also need to publish internal configuration updates. For this example it simple creates new updates to two items, "config1"and "config2" on a periodic basis (for simplicity both items share the same structure).
+
+### Config Broker (Broker)
+The **config broker** exposes two cross compartment methods, set_config() to publish and on_config() to subscribe. Internally it maintains the most recently published value of each item and the list of subscribers. Importantly it has no predefined knowledge of which items, publishers, or subscribes exist, and it's only level of trust is provided by the static sealed capabilities that only it can inspect.
+
+### Compartment[1-3] (Subscribers)
+These compartments provide the role of subscribers; Compartment 1 is allowed to receive "config1", Compartment 2 is allowed receive "config2", and Compartment 3 is allowed to receive both "config1" and "config2".
+
+### Validator
+This compartment provides a sandpit in which a compartment can validate a configuration update without exposing its own stack or context; It is prevented from making any heap allocations of its own. This is consistent with the principle that incoming data should be treated as being not only unsafe in content, but potentially even unsafe to parse. It exposes a single cross compartment method to validate the configuration struct used in the example; in practice each config item would need its own validation method. The validation used here is very basic, and implemented in a way to deliberately expose it to bound violations.
+
+## Threads in the example
+A number of thread are used to provide the dynamic behaviour. They are described in the context of the compartments where they start and execute their idle loops, although of course they also perform cross compartment calls. 
+
+A thread in the Config Source compartment periodically makes updates to both of the configuration items. A separate thread in Config Source compartment occasionally generates an invalid structure for "config1" which triggers a bounds violation in the validator.
+
+A thread in the Config Broker invokes the callbacks when a configuration item is changed. It uses a _futex_ to sleep until updates are available.  Decoupling the arrival of new data from its delivery to subscribers ensures that the thread which is making the updated (which may be part of a network stack) isn't blocked until all of the subscribers have processed it.  
```

Comment:
Note that this still means that subscribers *are* trusted for availability.  They can infinite loop in the callback and that will prevent any other notifications being delivered.  You could avoid this by exposing a read-only capability to a `uint32_t` that they then did a futex wait on (for C++, using `std::atomic<uint32_t>` and its `notify_*` / `wait` overloads is a nicer interface than using `futex_*` directly).  They could then call back in to query the new value.

You could actually avoid all cross-compartment calls if you handed everyone a read+cap object that is:

```c++
struct PublishedObject
{
  std::atomic<uint32_t> version;
  void *value;
};
```

They could then block on the futex changing and, when it did, read and `heap_claim_fast` the value.  The read function would look something like this:

```c++
bool wait(Timeout *t, PublishedObject &object, uint32_t oldVersion)
{
  object.wait(t, oldVersion);
  return object.load() != oldVersion;
}
int read(Timeout *t, PublishedObject &object, auto callback)
{
  void* value = object.value;
  if (int ret = heap_claim_fast(t, value); ret != 0))
  {
    return ret;
  }
  callback(value);
}
```

You'd then use `wait` to poll (blocking as determined with the timeout parameter) for changes and then use `read` to read the new value.  There's a little bit of fragility in `read`: it can't do any cross-compartment calls or the value may disappear out from under it.